File: bin/feyn-higgs-mssm.cc
#include <stdlib.h>

#include "TString.h"
#include "HiggsAnalysis/HiggsToTauTau/interface/FeynHiggsScan.h"

int main(int argc, char* argv[]) 
{
  // parse arguments
  if( argc < 6 ){
    std::cout << "Usage : " << argv[0] << " [type] [model] [process] [mA] [tanb] [option]       \n "
	      << " [option] : should consist of a key word and a value separated by a '='-sign. \n "
	      << "            keywords can be 'uncertainty' or 'model'. No spaces are allowed   \n "
	      << "            between keyword and value. Possible values for the options are:   \n "
	      << "            - uncertainty=mu- \n "
	      << "            - uncertainty=mu+ \n "
	      << "            - model=mhmax-7TeV\n "
	      << "            - model=mhmax-8TeV\n "
	      << "            - model=nomix-7TeV\n "
	      << "            - model=saeff-7TeV\n "
	      << "            - model=gluph-7TeV\n "
	      << "At the moment uncertainties are not fully supported. A set of examples is     \n "
	      << "given below:                                                                  \n "            
	      << " - " << argv[0] << " xs sm ggH 120 5 uncertainty=mu+    \n "
	      << " - " << argv[0] << " xs sm ggH 120 5 model=saeff-7TeV   \n ";
    return 0;
  }
  std::string model("mhmax-7TeV");
  if( argc > 6 ){
    if( std::string(argv[6]).find("model=")!=std::string::npos ){
      model = std::string(argv[6]).substr(std::string(argv[6]).find("=")+1);
    }
  }
  std::string uncertainty;
  if( argc > 6 ){
    if( std::string(argv[6]).find("uncertainty=")!=std::string::npos ){
      uncertainty+="_"+std::string(argv[6]).substr(std::string(argv[6]).find("=")+1);
    }
  }
  /*
    Implementation
  */
  if(atof(argv[5])<10){
    TString path = TString::Format("%s/src/HiggsAnalysis/HiggsToTauTau/data/feyn-higgs-mssm-scan-%s-fine.root", getenv("CMSSW_BASE"), model.c_str());
    //std::string path(getenv("CMSSW_BASE")+std::string("/src/HiggsAnalysis/HiggsToTauTau/data/feyn-higgs-mssm-scan-mhmax-fine.root"));
    FeynHiggsScan scan(path, TString::Format("mssm_scan%s", uncertainty.c_str()), 90, 1., 10., 96., 50., 1010.);
    std::cout << "tanb   = " << argv[5] << std::endl;
    std::cout << "mA     = " << argv[4] << std::endl;
    std::cout << "mH     = " << scan.mH(atof(argv[4]), atof(argv[5])) << std::endl;
    std::cout << "mh     = " << scan.mh(atof(argv[4]), atof(argv[5])) << std::endl;
    std::cout << "var    = " << argv[1] << "_" << argv[2] << "_" << argv[3] << std::endl;
    std::cout << "value  = " << scan.get(argv[3], argv[2], argv[1], atof(argv[4]), atof(argv[5])) << std::endl;
    return scan.get(argv[3], argv[2], argv[1], atof(argv[4]), atof(argv[5]));
  }
  else{
    TString path = TString::Format("%s/src/HiggsAnalysis/HiggsToTauTau/data/feyn-higgs-mssm-scan-%s.root", getenv("CMSSW_BASE"), model.c_str());
    //std::string path(getenv("CMSSW_BASE")+std::string("/src/HiggsAnalysis/HiggsToTauTau/data/feyn-higgs-mssm-scan-mhmax.root"));
    FeynHiggsScan scan(path, TString::Format("mssm_scan%s", uncertainty.c_str()), 70, 1., 71., 96., 50., 1010.);
    std::cout << "tanb   = " << argv[5] << std::endl;
    std::cout << "mA     = " << argv[4] << std::endl;
    std::cout << "mH     = " << scan.mH(atof(argv[4]), atof(argv[5])) << std::endl;
    std::cout << "mh     = " << scan.mh(atof(argv[4]), atof(argv[5])) << std::endl;
    std::cout << "var    = " << argv[1] << "_" << argv[2] << "_" << argv[3] << std::endl;
    std::cout << "value  = " << scan.get(argv[3], argv[2], argv[1], atof(argv[4]), atof(argv[5])) << std::endl;
    return scan.get(argv[3], argv[2], argv[1], atof(argv[4]), atof(argv[5]));
  }
}

File: python/tools/feyn_higgs_mssm.py
import ROOT
import os

class feyn_higgs_mssm:
    """
    Description:
    Class for determination of masses, cross-sections and branchingratios for a given higgs model.
    Arguments are mA, tanb, model, fmodel and period
      - model is from  'sm', 'mssm'.
      - fmodel is from 'mhmax', 'gluph', 'saeff'
      - period is from '7TeV', '8TeV'
    Allowed channels for get_xs are gg[h,H,A], qq[h,H,A]
    Allowed channels for get_br are [h,H,A]tt, [h,H,A]mm, [h,H,A]bb
    """
    def __init__(self, mA, tanb, model, fmodel, period):
        self.mA = mA
        self.tanb = tanb
        self.model = model
        if fmodel == '':
            fmodel = 'mhmax'
        path = os.getenv("CMSSW_BASE")+'/src/HiggsAnalysis/HiggsToTauTau/data/feyn-higgs-mssm-scan-'+fmodel+'-'+period
        if self.tanb < 10:
            self.scan = feyn_higgs_scan(path+'-fine.root', "mssm_scan", 90, 1, 10, 96, 50, 1010) 
        else:
            self.scan = feyn_higgs_scan(path+'.root', "mssm_scan", 70, 1, 71, 96, 50, 1010)
    def get_mh(self):
        mh = self.scan.mass("mh", self.mA, self.tanb)
        return mh
    def get_mH(self):
        mH = self.scan.mass("mH", self.mA, self.tanb)
        return mH
    def get_xs(self, channel):
        """
        This function returns the cross-section for a given production-channel.
        """
        xs = self.scan.get(channel, self.model, "xs",  self.mA, self.tanb)
        return xs
    def get_br(self, channel):
        """
        This function returns the branchingratio for a given decay-channel.
        """
        br = self.scan.get(channel, self.model, "br",  self.mA, self.tanb)
        return br

class feyn_higgs_scan:
    """
    Description:

    Basic class to determine branchingratios, cross-sections and masses of higgses.
    As input it takes the name of the rootfile and the TTree to be used as well as the
    parameters for the tanb and mA range.
    """
    def __init__(self, filename, treename, ntanb, mintanb, maxtanb, nmA, minmA, maxmA):
        self.mintanb = mintanb
        self.maxtanb = maxtanb
        self.steptanb = (maxtanb-mintanb)/float(ntanb)
        self.minmA = minmA
        self.maxmA = maxmA
        self.stepmA = (maxmA-minmA)/float(nmA)
        self.hproject = ROOT.TH2F("hproject", "hproject", nmA, minmA, maxmA, ntanb, mintanb, maxtanb)
        self.file = ROOT.TFile(filename, 'read')
        if self.file.IsZombie():
            print "Error: Could not open file"
        self.tree = self.file.Get(treename)

    def project(self, var):
        self.hproject.Reset()
        for idx in range(0, self.tree.GetEntries()):
            self.tree.GetEvent(idx)
            self.hproject.Fill(self.tree.mA + self.stepmA/2., self.tree.tanb+self.steptanb/2., getattr(self.tree,var))

    def mass(self, var, mA, tanb):
        """ Extraction of mass from TTree """
        mass = 0.
        for idx in range(0,self.tree.GetEntries()):
            self.tree.GetEvent(idx)
            if self.tree.mA == mA and self.tree.tanb == tanb:
                mass = getattr(self.tree,var)
                break
        return mass

    def get(self, variable, model, type, mA, tanb):
        """ extract value of 'variable' from rootfiles """
        self.project(type+'_'+model+'_'+variable)
        if self.minmA <= mA and mA < self.maxmA and self.mintanb <= tanb and tanb < self.maxtanb:
            return self.hproject.GetBinContent(self.hproject.FindBin(mA, tanb))
        else:
            return -999
        
